### ISSUE TITLE
2 main problems with UFO format

### ISSUE THREAD

**Issue by aminabedi68** [fontforge/fontforge#4005]
hi
i am creating variable fonts with fontforge and fontmake, so i have to use UFO as output format of fontforge and the input format of fontmake, but there are 2 problems:

**1- wrong kerning pair codes writing method in features.fea file** ,fontforge writing these codes like this:
```
pos \uni062F <-122 0 -122 0> \uniFB7C  <-122 0 -122 0>;
```
but it's not working and the right code is this:

```
pos \uni062F \uniFB7C <-122 0 -122 0>;
```
I'm pretty sure fontforge writing code is wrong (looks corrupted even when you try to open with fontforge).

**2- issue about ligature mark suffix cause mark interpolation problems** ,please read this issue:
https://github.com/googlefonts/fontmake/issues/585#issuecomment-547480446

these are two main problems i know about creating variable fonts using opensource softwares.

**Comment by skef** [comment]
Would it be possible to supply some sample files and export instructions that exhibit these problems? I believe that pair kerning and feature files works in at least some cases, so it would help to have an example.

**Comment by aminabedi68** [comment]
hi @skef 
you can test on my font, <a href="https://github.com/aminabedi68/Estedad/blob/master/Source/variable/5-Estedad-V3.0%20-%20Black.sfd">Estedad </a> 
to see first issue you can simply open the font and generate as UFO, then open the UFO with fontforge again, you will see different kerning pairs table in some GPOS kerning tables like kernHorizontalKerninginArabiclookup2 if you generate a font with this UFO, kernings are not the same

fontforge file:
![fontforge_2019-11-05_22-22-45](https://user-images.githubusercontent.com/25493297/68236758-036fd880-001b-11ea-86c5-79762e71c53c.png)




generated UFO file:
![fontforge_2019-11-05_22-23-14](https://user-images.githubusercontent.com/25493297/68236792-14b8e500-001b-11ea-8907-fc07838353b1.png)
last two columns are the problem




for the second one just open features.fea in generated UFO folder with a text editor, you can see ligatures mark has not any suffix, for example line 1408 to 1410 is this:
```
  pos ligature \uniFEFB <anchor 830 0> mark @Anchor2 
    ligComponent
      <anchor 241 0> mark @Anchor2 ;
```

**Comment by skef** [comment]
The kerning pair problem seems to be a simple misinterpretation of the spec that has gone unnoticed since the start of feature file support. I have a fix that I'll try to verify a bit more soon. 

The UFO anchor situation is more of a mess. I could probably fix it if I knew exactly what the behavior should be, but I don't so I'll have to try to get that discussion going.

**Comment by aminabedi68** [comment]
thank you @skef
UFO format is very important if we want to work with several softwares and unfortunately 
the UFO format of fontforge on export is not the same explained <a href="http://unifiedfontobject.org/versions/ufo3/index.html">here </a>, it includes features.fea that ot features already applied on it, i mean kernings, marks, etc. so the main problem is how to write all information of a font on fully uncompressed UFO.

**Comment by skef** [comment]
The UFO specification doesn't address your second issue, as far as I know; that's a convention that has become a quasi-standard due to its solving a problem people need solved. 

The general problem here is that the UFO format now supports anchor points in glifs and also anchor points in feature files. They need to be in the latter and the spec doesn't indicate they shouldn't appear in both places. That means that on import you need to be careful about deciding which points are really the same point, and FontForge isn't doing that now (see #3986). That basic logic needs to be figured out before we start doing more subtle things like implementing this other convention.

**Comment by aminabedi68** [comment]
yes @skef ,you right, but the problem about exported UFO(exactly about ot features like kerning,marks,... that written in features.fea file) is that the other softwares (like fontmake(see: https://github.com/googlefonts/fontmake/issues/585#issuecomment-555436859), fontlab,...) think the font already compiled and it cause many other problems. if we create a font and export UFO, we should be able to generate the same font with other softwares using the same UFO(this is what i understood, if you have more information please guid me)

**Comment by skef** [comment]
Relative to problem 2: I've looked at the `fontmake` issue and this one and at this point I *sort of* understand how FontForge is doing something wrong but I neither understand specifically how to fix it nor what use case fixing it would help. If this is just a matter of anchor name choice then it seems as though one could just pick different names. If they have to match up in a specific way I would need to know what that way is in order to help.
